File: app/src/main/java/me/ameriod/lib/mvp/app/models/Person.java
package me.ameriod.lib.mvp.app.models;

import android.net.Uri;
import android.os.Parcelable;
import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;
import com.google.gson.Gson;
import com.google.gson.TypeAdapter;
import com.google.gson.annotations.SerializedName;

import java.util.List;

/**
 * "starships": [
 * "http://swapi.co/api/starships/12/",
 * ...
 * ],
 * "url": "http://swapi.co/api/people/1/",
 * "vehicles": [
 * "http://swapi.co/api/vehicles/14/"
 * ...
 * ]
 * }
 */
@AutoValue
public abstract class Person implements Parcelable {

    public static TypeAdapter<Person> typeAdapter(Gson gson) {
        return new AutoValue_Person.GsonTypeAdapter(gson);
    }

    @Nullable
    @SerializedName("birth_year")
    public abstract String birthYear();

    @Nullable
    @SerializedName("eye_color")
    public abstract String eyeColor();

    @Nullable
    @SerializedName("films")
    public abstract List<String> films();

    @Nullable
    @SerializedName("gender")
    public abstract String gender();

    @Nullable
    @SerializedName("height")
    public abstract String height();

    @Nullable
    @SerializedName("homeworld")
    public abstract String homeworld();

    @Nullable
    @SerializedName("mass")
    public abstract String mass();

    @Nullable
    @SerializedName("name")
    public abstract String name();

    @Nullable
    @SerializedName("hair_color")
    public abstract String hairColor();

    @Nullable
    @SerializedName("skin_color")
    public abstract String skinColor();

    @Nullable
    @SerializedName("species")
    public abstract List<String> species();

    @Nullable
    @SerializedName("starships")
    public abstract List<String> starships();

    @Nullable
    @SerializedName("url")
    public abstract String url();

    @Nullable
    @SerializedName("vehicles")
    public abstract List<String> vehicles();

    public final String getSpecies() {
        StringBuilder builder = new StringBuilder();
        for (int i = 0, size = species().size(); i < size; i++) {
            builder.append(species().get(i)).append(" ");
        }
        return builder.toString();
    }

    public final String getId() {
        return Uri.parse(url()).getLastPathSegment();
    }
}

File: app/src/main/java/me/ameriod/lib/mvp/app/models/Person.kt
package me.ameriod.lib.mvp.app.models

import android.net.Uri
import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName

data class Person(@SerializedName("birth_year")
                  val birthYear: String?,
                  @SerializedName("eye_color")
                  val eyeColor: String?,
                  @SerializedName("films")
                  val films: List<String>?,
                  @SerializedName("gender")
                  val gender: String?,
                  @SerializedName("height")
                  val height: String?,
                  @SerializedName("homeworld")
                  val homeworld: String?,
                  @SerializedName("mass")
                  val mass: String?,
                  @SerializedName("name")
                  val name: String?,
                  @SerializedName("hair_color")
                  val hairColor: String?,
                  @SerializedName("skin_color")
                  val skinColor: String?,
                  @SerializedName("species")
                  val species: List<String>?,
                  @SerializedName("starships")
                  val starships: List<String>?,
                  @SerializedName("url")
                  val url: String?,
                  @SerializedName("vehicles")
                  val vehicles: List<String>?) : Parcelable {

    fun getSpecies(): String {
        val builder = StringBuilder()
        var i = 0
        val size = species!!.size
        while (i < size) {
            builder.append(species[i]).append(" ")
            i++
        }
        return builder.toString()
    }

    fun getVehicles(): String {
        val builder = StringBuilder()
        var i = 0
        val size = vehicles!!.size
        while (i < size) {
            builder.append(vehicles[i]).append(" ")
            i++
        }
        return builder.toString()
    }

    fun getStarShips(): String {
        val builder = StringBuilder()
        var i = 0
        val size = starships!!.size
        while (i < size) {
            builder.append(starships[i]).append(" ")
            i++
        }
        return builder.toString()
    }

    fun getFilms(): String {
        val builder = StringBuilder()
        var i = 0
        val size = films!!.size
        while (i < size) {
            builder.append(films[i]).append(" ")
            i++
        }
        return builder.toString()
    }

    fun getId(): String? {
        return Uri.parse(url).lastPathSegment
    }

    constructor(source: Parcel) : this(
            source.readString(),
            source.readString(),
            source.createStringArrayList(),
            source.readString(),
            source.readString(),
            source.readString(),
            source.readString(),
            source.readString(),
            source.readString(),
            source.readString(),
            source.createStringArrayList(),
            source.createStringArrayList(),
            source.readString(),
            source.createStringArrayList()
    )

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeString(birthYear)
        writeString(eyeColor)
        writeStringList(films)
        writeString(gender)
        writeString(height)
        writeString(homeworld)
        writeString(mass)
        writeString(name)
        writeString(hairColor)
        writeString(skinColor)
        writeStringList(species)
        writeStringList(starships)
        writeString(url)
        writeStringList(vehicles)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<Person> = object : Parcelable.Creator<Person> {
            override fun createFromParcel(source: Parcel): Person = Person(source)
            override fun newArray(size: Int): Array<Person?> = arrayOfNulls(size)
        }
    }
}